### PR TITLE
TypingDNA functions

### PR DESCRIPTION
**Pull request by Gayarohanangajanan**
initial requests

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    <version>1.0.0-SNAPSHOT</version>
+    <name>Conditional Authentication - Custom Functions</name>
+    <packaging>bundle</packaging>
+
+
+    <dependencies>
+        <dependency>
+            <groupId>commons-lang.wso2</groupId>
+            <artifactId>commons-lang</artifactId>
+            <version>2.6.0.wso2v1</version>
```
[the remark this answers is not in the document]
Comment (Gayarohanangajanan) [reply pom.xml]: extracted

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/SaveUserInTypingDNAFunctionImpl.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            StringBuilder res = new StringBuilder();
+            String line;
+            while ((line = rd.readLine()) != null) {
+                res.append(line);
+                res.append('\r');
+            }
+            rd.close();
+
+            // Response from TypingDNA
+            log.debug(res.toString());
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/SaveUserInTypingDNAFunctionImpl.java:93]: You need to check whether the debug mode is enabled. Also, format the log in a proper manner.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/TypingDNAConfigImpl.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Override
+    public Map<String, String> getPropertyDescriptionMapping() {
+
+        Map<String, String> mapping = new HashMap<>();
+
+        mapping.put(ENABLE, "Enable TypingDNA Authentication");
+        mapping.put(USERNAME, "Typing DNA API Key");
+        mapping.put(CREDENTIAL, "Typing DNA API Secret");
+        mapping.put(ADVANCE_MODE_ENABLED, "Enable if your typingDNA account is pro");
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/TypingDNAConfigImpl.java:100]: Revisit the wordings.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/utils.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsServletRequest;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * class that implements methods to extract User details and
+ * typing patterns from the context
+ */
+public class utils {
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/utils.java:35]: The class name should start with a capital letter. It's the naming convention we follow.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/SaveUserInTypingDNAFunctionImpl.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        // Getting connector configurations.
+        String APIKey = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.USERNAME, tenantDomain);
+        String APISecret = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.CREDENTIAL, tenantDomain);
+        String advanced = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.ADVANCE_MODE_ENABLED, tenantDomain);
+        String region = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.REGION, tenantDomain);
+
+        String userID = getUserID(username, tenantDomain);
+
+        if (!(typingPattern == null) && !typingPattern.equals(Constants.NULL)) {
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/SaveUserInTypingDNAFunctionImpl.java:63]:
```suggestion
        if (StringUtils.isNotBlank(typingPattern)  && !StringUtils.equalsIgnoreCase(Constants.NULL, typingPattern)) {
```

Comment (ashensw) [reply src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/SaveUserInTypingDNAFunctionImpl.java:63]: `org.apache.commons.lang.StringUtils`

File: resources/typing-dna/plugins/basicauth-extensions.jsp [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<%@ page import="java.io.File" %>
+
+<%
+    File tpattern = new File(getServletContext().getRealPath("plugins/typing-dna.jsp"));
+    if (tpattern.exists() && !isIdentifierFirstLogin(inputType)) {
+%>
+        <jsp:include page="plugins/typing-dna.jsp"/>
+<% }  %>
```
Comment (ashensw) [inline resources/typing-dna/plugins/basicauth-extensions.jsp:8]: Add a new line

File: resources/typing-dna/plugins/typing-dna.jsp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+<% } %>
+<script>
+
+var tdna = new TypingDNA();
+tdna.addTarget("usernameUserInput");
+tdna.addTarget("password");
+
+$(document).ready(function(){
+        //document.getElementById("loginForm").onsubmit = function() {getTypingPatterns()};
+        //document.getElementById("loginForm").setAttribute("onsubmit", "getTypingPatterns()");
```
Comment (ashensw) [inline resources/typing-dna/plugins/typing-dna.jsp:20]: Do we need these commented lines?

File: resources/typing-dna/plugins/typing-dna.jsp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        document.getElementById("loginForm").addEventListener("submit",getTypingPatterns);
+        document.getElementById("usernameUserInput").setAttribute("autocomplete", "off");
+});
+
+function getTypingPatterns() {
+
+            var tp1 = tdna.getTypingPattern({type:1});
+
+            //adding a new field to form
```
Comment (ashensw) [inline resources/typing-dna/plugins/typing-dna.jsp:29]: Let's follow the proper format for the comments.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunction.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsAuthenticationContext;
+
+import java.util.Map;
+
+@FunctionalInterface
+public interface VerifyUserWithTypingDNAFunction {
+
+    /**
+     * Custom adaptive function for VERIFY
+     * users typing pattern in TypingDNA.
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunction.java:30]: Add the param descriptions.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+public class VerifyUserWithTypingDNAFunctionImpl implements VerifyUserWithTypingDNAFunction {
+
+    private static final Log log = LogFactory.getLog(VerifyUserWithTypingDNAFunctionImpl.class);
+
+    /**
+     * Function to send verify request to typingDNA APIs.
+     *
+     * @param context       Context from authentication flow.
+     * @param eventHandlers
+     * @throws TypingDNAAuthenticatorException
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java:68]: Complete the comment descriptions.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            String username = user.getWrapped().getUserName();
+            String tenantDomain = user.getWrapped().getTenantDomain();
+            String typingPattern = Utils.getTypingPattern(context);
+
+            AtomicBoolean result = new AtomicBoolean(false);
+
+            // Getting connector configurations.
+            String APIKey = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.USERNAME, tenantDomain);
+            String APISecret = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.CREDENTIAL, tenantDomain);
+            String advanced = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.ADVANCE_MODE_ENABLED, tenantDomain);
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java:84]: Shall we use boolean for this and rename it as `isAdvancedModeEnabled`?

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+            String region = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.REGION, tenantDomain);
+            String Enabled = CommonUtils.getConnectorConfig(TypingDNAConfigImpl.ENABLE, tenantDomain);
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java:86]: Shall we use boolean here as well. You can use Boolean.parseBoolean(). Change the usages accordingly.

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+                        String line;
+                        while ((line = rd.readLine()) != null) {
+                            res.append(line);
+                            res.append('\r');
+                        }
+                        rd.close();
+
+                        // Response from TypingDNA.
+                        if (log.isDebugEnabled()) {
+                            log.debug("Response from TypingDNA for the user  " + username + ":  " + res.toString());
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/VerifyUserWithTypingDNAFunctionImpl.java:129]:
```suggestion
                            log.debug("Response from TypingDNA for the user:  " + username + ":  " + res.toString());
```

File: src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/internal/TypingDNAFunctionsServiceComponent.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+import org.wso2.carbon.identity.conditional.auth.typingdna.TypingDNAConfigImpl;
+import org.wso2.carbon.identity.conditional.auth.typingdna.VerifyUserWithTypingDNAFunction;
+import org.wso2.carbon.identity.conditional.auth.typingdna.VerifyUserWithTypingDNAFunctionImpl;
+import org.wso2.carbon.identity.governance.IdentityGovernanceService;
+import org.wso2.carbon.identity.governance.common.IdentityConnectorConfig;
+import org.wso2.carbon.registry.core.service.RegistryService;
+import org.wso2.carbon.user.core.service.RealmService;
+
+/**
+ * OSGi declarative services component which handled registration and un-registration of TypingDNA related
```
Comment (ashensw) [inline src/main/java/org/wso2/carbon/identity/conditional/auth/typingdna/internal/TypingDNAFunctionsServiceComponent.java:43]: Add this as the class comment.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+    <properties>
+        <carbon.kernel.version>4.5.1</carbon.kernel.version>
```
Comment (ashensw) [inline pom.xml:222]: Use the latest kernel GA version. Update other versions as well.

File: pom.xml [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+    <properties>
+        <carbon.kernel.version>5.2.13</carbon.kernel.version>
```
Comment (ashensw) [inline pom.xml:222]:
```suggestion
        <carbon.kernel.version>4.6.1</carbon.kernel.version>
```

File: pom.xml [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+        <carbon.utils.version>4.6.1</carbon.utils.version>
+        <carbon.user.core.version>4.6.1</carbon.user.core.version>
```
Comment (ashensw) [inline pom.xml:224]: You can use `<carbon.kernel.version>` for these two as well. No need to define this separately.

File: pom.xml [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+        <identity.governance.version>1.4.95</identity.governance.version>
+        <com.googlecode.json-simple.wso2.version>1.1.wso2v1</com.googlecode.json-simple.wso2.version>
+        <json-smart.version>2.3</json-smart.version>
+        <carbon.identity.framework.version>5.14.97</carbon.identity.framework.version>
```
Comment (ashensw) [inline pom.xml:228]:
```suggestion
        <carbon.identity.framework.version>5.19.58</carbon.identity.framework.version>
```

File: pom.xml [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+        <imp.pkg.version.javax.servlet>[2.5.0, 3.0.0)</imp.pkg.version.javax.servlet>
+        <carbon.identity.framework.imp.pkg.version.range>[5.14.97, 6.0.0)
+        </carbon.identity.framework.imp.pkg.version.range>
+        <osgi.framework.imp.pkg.version.range>[1.7.0, 2.0.0)</osgi.framework.imp.pkg.version.range>
+        <osgi.service.component.imp.pkg.version.range>[1.2.0, 2.0.0)</osgi.service.component.imp.pkg.version.range>
+        <carbon.identity.conditional.auth.functions.imp.pkg.version.range>[1.1.6, 2.0.0)
+        </carbon.identity.conditional.auth.functions.imp.pkg.version.range>
+        <carbon.user.api.imp.pkg.version.range>[1.0.1, 2.0.0)</carbon.user.api.imp.pkg.version.range>
+        <identity.governance.imp.pkg.version.range>[1.4.90, 2.0.0)</identity.governance.imp.pkg.version.range>
+        <carbon.identity.application.imp.pkg.version>5.19.54</carbon.identity.application.imp.pkg.version>
```
Comment (ashensw) [inline pom.xml:245]: You can use the carbon.identity.framework.version for this as well?

**Review by ashensw — CHANGES_REQUESTED** [review]

**Review by ashensw — APPROVED** [review]

**Pull request merged**